server/src/test/java/e2e/RetentionTest.java

```diff
@@ -0,0 +1,112 @@
+public class RetentionTest {
+
+    public static final String RETENTION_PARENT_EVENT = "retention-parent-event";
+    public static final String RETENTION_CHILD_EVENT = "retention-child-event";
+
+    private WorkflowVerifier verifier;
+    private LittleHorseBlockingStub lhClient;
+
+    @LHWorkflow("child-thread-retention-test")
+    private Workflow threadRetentionWorkflow;
+
+    @LHWorkflow("wf-run-retention-test")
+    private Workflow wfRunRetentionWorkflow;
+
+    @LHWorkflow("child-thread-retention-test")
+    public Workflow buildRetentionWorkflow() {
+        return Workflow.newWorkflow("retention-test", wf -> {
+            wf.spawnThread(
+                    child -> {
+                        child.withRetentionPolicy(ThreadRetentionPolicy.newBuilder()
+                                .setSecondsAfterThreadTermination(1)
+                                .build());
+                        child.waitForEvent(RETENTION_CHILD_EVENT);
+                    },
+                    "child",
+                    null);
+
+            // Wait twice so we have control and can inspect the WfRun twice.
+            wf.waitForEvent(RETENTION_PARENT_EVENT);
+            wf.waitForEvent(RETENTION_PARENT_EVENT);
+
+            // Ensure that the threadRunNumber for this thread is 2, not 1
+            wf.spawnThread(child -> {}, "second-child", null);
+        });
+    }
+
+    @LHWorkflow("wf-run-retention-test")
+    public Workflow buildWfRunRetentionWorkflow() {
+        Workflow out = Workflow.newWorkflow("wf-run-retention-test", wf -> {});
+        out.withRetentionPolicy(WorkflowRetentionPolicy.newBuilder()
+                .setSecondsAfterWfTermination(1)
+                .build());
+        return out;
+    }
+
+    @Test
+    void shouldDeleteWfRunAfterRetentionExpires() {
+        WfRunId id = verifier.prepareRun(wfRunRetentionWorkflow)
+                .waitForStatus(LHStatus.COMPLETED)
+                .start();
+
+        Awaitility.await().atMost(Duration.ofSeconds(4)).until(() -> {
+            return isNotFound((client) -> {
+                client.getWfRun(id);
+                return true;
+            });
+        });
+    }
+
+    private boolean isNotFound(Function<LittleHorseBlockingStub, Boolean> callable) {
+        StatusRuntimeException caught = null;
+        try {
+            callable.apply(lhClient);
+        } catch (StatusRuntimeException exn) {
+            if (LHTestExceptionUtil.isNotFoundException(exn)) {
+                caught = exn;
+            }
+        }
+        return caught != null;
+    }
+
+    @Test
+    void childThreadShouldGcAfterOneSecond() {
+        verifier.prepareRun(threadRetentionWorkflow)
+                .thenVerifyWfRun(wfRun -> {
+                    Assertions.assertThat(wfRun.getThreadRunsCount()).isEqualTo(2);
+                })
+                .thenSendExternalEventJsonContent(RETENTION_CHILD_EVENT, Map.of())
+                .thenSleep(1000)
```

Comment:
This seems like the old e2e framework. 

I don't really recommend adding this as an option to develop e2e in our context. Instead, we should wait for the server to handle the external events, using GRPC calls to query the external events.